SPARQL -> FQL translator refactor

dpetran opened the pull request:
This is a rewrite of the SPARQL translator that solves several issues, and should make adding support for more SPARQL syntax easier.

Specifically fixes :union patterns in where clauses, subject iri parsing, and REGEX functions.

fixes https://github.com/fluree/core/issues/90

dpetran committed:
clean parse tree

remove unneeded structure

dpetran committed:
remove constant condition

dpetran committed:
fix sparql union queries

dpetran committed:
factor out sparql select-clause handling

dpetran committed:
map, apply concat -> mapcat

dpetran committed:
reorganize BuiltInCall grammar

This groups the other rules together at the top and then alphabetizes the func names.

BuiltInCall ::= Aggregate
| ExistsFunc
| NotExistsFunc
| RegexExpression
| StrReplaceExpression
| SubstringExpression
| '<FUNC>'

dpetran committed:
rewrite sparql translator

This provides a much simpler architecture for adding additional SPARQL support.

A few external changes were necessary:
1) I changed the FQL schema for :order-by clauses to not strictly require lists, but any
sequential collection. This is fine, as the only place we use it just destructures the
values, we are in no way depending on list-specific behaviors.
2) The SPARQL translator tests for :bind patterns were requiring FQL that didn't
actually work. The syntax changed when we changed the where syntax, and the translated
:bind syntax was never executed.
3) :group-by clauses can contain multiple variables or expressions, and it was simpler
for this implementation to just always return a sequence, regardless if there was one or
many conditions.

I did modify the grammar a bit, only by factoring out the :Func term to make
:BuiltInCall processing simpler.

This is a WIP commit, I plan to clean it up in the next one after I address the final
failing test.

dpetran committed:
fix order-by clause translation

We _can_ (and actually require) a multicardinality order-by clause.

:order-by ["desc" "?var"] - valid fql, but doesn't work
:order-by [["desc" "?var]] - valid fql, works

dpetran committed:
factor out Func from BuiltInCall rule

This makes it a lot easier to handle the structure correctly.

dpetran committed:
hide GraphTerm parse rule

Since it's just a union of other types and imposes no structure of its own, it's safe to
hide.

dpetran committed:
remove unneeded parse node annotations

This data is already encoded in the node's tag, and is redundant.

dpetran committed:
introduce new parse rule for ExplicitOrderCondition

This makes handling the order clause a lot simpler.

dpetran committed:
remove node alias

Unneeded structure.

dpetran committed:
switch to new sparql translator implementation

dpetran committed:
remove experimental grammar

dpetran committed:
add test for subject iris

dpetran committed:
add support for :RegexExpressions

dpetran committed:
fix cljs impl

dpetran committed:
minimize diff

This line accidentally got moved, so I'm moving it back to keep the grammar modification
history minimal.

dpetran committed:
add support for multiple values per var in VALUES pattern

In SPARQL, a VALUES pattern can appear inside a WHERE clause alongside other
patterns. In FQL :values is its own clause. This mismatch makes translating directly
awkward, since we need to insert our translated pattern at a different point in the
parse tree than where we are at translation time.

We got around it by translating it into a :bind pattern. However, a :bind pattern can
only bind a single value to a variable, so we limited the scope of a VALUES pattern to
only a single value.

This gets around it by "tagging" the translated value pattern and then after the whole
parsing is done, moving it to the correct place in the tree: see `format-values`. It's
awkward, but it gives much more flexibility to the user.

dpetran committed:
comment out incorrect property path translation tests

These tests expect a translation that does not work in FQL. Property paths are not
supported at all in FQL, and until they are these tests provide no value.

zonotope commented on src/clj/fluree/db/query/fql/syntax.cljc:
```diff
@@ -131,12 +131,12 @@
     ::ordering        [:orn {:error/message "Ordering must be a var or two-tuple formatted ['ASC' or 'DESC', var]"}
                        [:scalar ::var]
-                       [:vector [:and list?
+                       [:vector [:and sequential?
```
I think the ordering must be an explicit list to differentiate it from our syntactic usage of vectors for things like groups of variables.

zonotope commented on test/fluree/db/query/json_ld_compound_test.clj:
```diff
@@ -118,7 +118,7 @@
                                  :where   {:schema/name ?name
                                            :ex/favNums  ?favNums}
-                                 :orderBy (desc ?favNums)})
+                                 :orderBy [(desc ?favNums)]})
```
Does this change require ordering clauses to be wrapped in a vector now?

zonotope replied:
Did you change it? If so, could you change this test back too?

cap10morgan replied:
This still needs to be done, it looks like?

dpetran committed:
support VALUES patterns in where clause

This is a basic implementation of the SPARQL VALUES pattern type.

dpetran committed:
support VALUES patterns in an order-agnostic way

Three approaches considered:

1) just document that :values needs to appear first
2) discard solutions that don't match the given solution
3) during the parse step, sort the patterns so :values is first

I don't like #1, though that's the most expedient. #2 is a bit wasteful as it forces us
to generate useless solutions. #3 is doable, and may be a minimal requirement in basic
query planning.

This implements solution #2.

dpetran committed:
add values test for nested patterns

dpetran committed:
maintain list ordering syntax

changing ::ordering from `list?` to `sequential?` was not a backwards compatible change,
and unnecessary as it turns out. This reverts that change.

dpetran committed:
add support for VALUES patterns and VALUES clauses

FQL now supports :values patterns so we can use that directly instead of translating to
a :values clause.

dpetran committed:
SUBSTR and REPLACE expresssion translations

dpetran committed:
Throw errors for unsupported EXISTS and NOT EXISTS filters

dpetran committed:
add support for complex multiplicative expressions

dpetran committed:
handle Func translation on a case-by-case basis

The tricky part is deciding when to `literal-quote` one or more of the args. Since that
information does not exist in the parse tree at the point where :Func needs parsing, we
need to figure out whether to quote based on the signature of the function, according to
the SPARQL spec: https://www.w3.org/TR/sparql11-query/#SparqlOps and the SPARQL grammar.

This way is more repetitive but more robust. Once we've got all the functions covered we
can search for abstractions to reduce verbosity.

dpetran committed:
add support for BOUND

JaceRockman committed:
Add separate test for funcs

JaceRockman committed:
Alphabetize the lists of funcs

dpetran committed:
support SPARQL literal datatype and language tags

I'm not sure where the LANGTAG regex came from, but it didn't match the one in the spec
- it required the hyphen (-) between two-part lang tags of the style `@en-us`, but
disallowed the simple verson `@en`. Now it correctly accepts both.

dpetran committed:
confirm SPARQL comments are ignored

JaceRockman committed:
add support for BNODE

JaceRockman committed:
add support for COALESCE

JaceRockman committed:
fix implementation of BNODE

JaceRockman committed:
add support for CONTAINS

JaceRockman committed:
add support for DATATYPE

JaceRockman committed:
add support for DAY

JaceRockman committed:
add support for ENCODE_FOR_URI

JaceRockman committed:
add support for FLOOR

JaceRockman committed:
add support for HOURS

JaceRockman committed:
add support for IF

JaceRockman committed:
add support for IRI

JaceRockman committed:
add support for LANG

JaceRockman committed:
add support for LANGMATCHES

JaceRockman committed:
add support for LCASE

JaceRockman committed:
add support for MD5

JaceRockman committed:
add support for MINUTES

JaceRockman committed:
add support for MONTH

JaceRockman committed:
add support for NOW

JaceRockman committed:
add support for RAND

JaceRockman committed:
add support for ROUND

JaceRockman committed:
add support for SECONDS

JaceRockman committed:
add support for SHA1

JaceRockman committed:
add support for SHA256

JaceRockman committed:
add tests for SHA512

JaceRockman committed:
add support for STR

JaceRockman committed:
add support for STRAFTER and STRBEFORE

JaceRockman committed:
add support for STRDT

JaceRockman committed:
add support for STRENDS

JaceRockman committed:
add support for STRLANG

JaceRockman committed:
add support for STRLEN and STRENDS

JaceRockman committed:
add support for STRUUID

JaceRockman committed:
add support for TIMEZONE

JaceRockman committed:
add support for TZ

JaceRockman committed:
add support for UCASE

JaceRockman committed:
add support for URI

JaceRockman committed:
add support for UUID

JaceRockman committed:
add support for YEAR

JaceRockman committed:
add support for isBLANK

JaceRockman committed:
add support for isIRI

JaceRockman committed:
add support for isLITERAL

JaceRockman committed:
add support for isNUMERIC

JaceRockman committed:
add support for isURI

JaceRockman committed:
add support for sameTerm

dpetran committed:
correctly handle multiple expressions joined logically

dpetran committed:
add error for SERVICE patterns

JaceRockman committed:
add support for FROM NAMED

JaceRockman committed:
Correct grammar comment in DatasetClause parse-rule

JaceRockman committed:
add support for GRAPH

dpetran committed:
add support for EXISTS and NOT EXISTS patterns

dpetran committed:
use existing helper function instead of destructuring

dpetran committed:
add explanations to test cases

dpetran committed:
use parse-where-clause function

dpetran committed:
allow SPARQL to use `a` as an alias for @type

The grammar wasn't expecting whitespace between the 'a' and the prev/next term.

dpetran committed:
correctly filter solutions across ledgers and with meta

Change 1: handle multiple inline values correctly.

Instead of checking that _every_ inline solution equal the solution matches, we needed
to _filter_ the inline solutions by whether they match the solution. If any matches, the
values can contribute their solutions to the solution chan.

This case allows multiple values to be handled correctly, as seen in the "pattern"
"single var" test.

Change 2: do not consider ::sids while checking match equality

When dealing with federated queries the ::sids map may be different for the same
value. We now dissoc the ::sids key from the solution matches before checking equality.

This is verified by the "federated" test case

Change 3: only consider :lang ::meta while checking equality

Two identical strings with different language tags are different values. However the :i
order or whether it comes from a reasoned flake do not matter for equality checks. We
now select only the :lang ::meta key, if ::meta exists in the solution match.

Language tags were not parsed, so I've added parsing support for that.

dpetran committed:
use public match api instead of removing unwanted match elements

In the future the contents of a match may change, so instead of removing specific
unwanted elements we now extract only those elements of a match that establish its
identity for equality checking.

dpetran committed:
fix typo

dpetran committed:
add description of sparql function translation

dpetran committed:
add support for `in` expression

Semantics and syntax from SPARQL IN expression.

cap10morgan commented on test/fluree/db/query/sparql_test.cljc:
```diff
@@ -140,6 +140,20 @@
+          {:keys [where]} (sparql/->fql query)]
+      (is (= [{"@id" "?s", "@type" "?t"}
+              {"@id" "?s", "schema:name" "?name"}
+              [:filter ["(regex ?name \"^Jon\" \"i\")"]]]
```
What is this testing? Just that REGEX filters are translated correctly? Might be good to add a `testing` annotation (or similar) to that effect. Looks like the `psm` prefix is unused, is that intentional?

cap10morgan commented on test/fluree/db/query/sparql_test.cljc:
```diff
@@ -233,14 +247,14 @@
                    ORDER BY DESC(?favNums)"
             {:keys [orderBy]} (sparql/->fql query)]
-        (is (= ["desc" "?favNums"]
+        (is (= [["desc" "?favNums"]]
```
Looks like these need to be changed back to single-dimensional vectors too?

cap10morgan commented on test/fluree/db/query/sparql_test.cljc:
```diff
@@ -254,7 +268,7 @@
                  GROUP BY ?e HAVING(SUM(?favNums) > 1000)"
           {:keys [groupBy having]} (sparql/->fql query)]
-      (is (= "?e"
+      (is (= ["?e"]
```
Needs to be reverted too?

dpetran replied:
Same deal here, the translator will always output a vector for ORDER BY, even though FQL accepts it both ways.

cap10morgan commented on test/fluree/db/query/sparql_test.cljc:
```diff
@@ -296,6 +311,24 @@
+SELECT ?p ?o
+FROM <cookbook/base>
+WHERE {
+    ex:andrew ?p ?o. # Should provide just the triples related to andrew
```
This comment might be more useful in a `testing` annotation?

dpetran replied:
This was a direct test case from the ticket, I think I'll add it in with the `simple-triple` expectations.

cap10morgan reviewed (commented):
Looks like there are still some outstanding TODOs in here

dpetran committed:
orderBy clause accepts single-cardinality values

dpetran committed:
move subject iri test to where translation test suite

dpetran committed:
remove unused prefix and add test explanation

dpetran committed:
translate SPARQL EXISTS/NOT EXISTS filters into FQL

dpetran committed:
SPARQL IN support

dpetran committed:
fix expression quoting

The :Expression parser was turning everything into a string, which works in some cases
but introduces unwanted quoting in others. By deferring stringification until values
need to be incorporated into an actual string expression we avoid the unwanted quoting.

Do note that all the numeric functions now correctly expect numeric values instead of
strings.

dpetran committed:
SPARQL NOT IN support

zonotope commented on src/clj/fluree/db/query/sparql/translator.cljc:
```diff
@@ -0,0 +1,489 @@
+    (str "\"" x "\"")
+    x))
+
+;; take a term rule and return a value
```
This comment should go into the docstring of the `parse-term` multimethod.

zonotope commented on src/clj/fluree/db/query/sparql/translator.cljc:
```diff
@@ -0,0 +1,489 @@
+
+;; take a term rule and return a value
+(defmulti parse-term (fn [[tag & _]] tag))
+;; take a rule return a sequence of entry tuples [[k v]...]
```
This comment should go into the docstring of the `parse-rule` multimethod

dpetran committed:
add docstring to translator multimethods

dpetran committed:
correctly assign langString datatype to language tagged strings

We were not parsing language tagged strings to langString datatypes correctly in where
patterns or in values patterns, nor inferring a langString datatype in the presence of a
language tag.

I had to update the test expectation because now the language tagged string no longer
conforms to the `sh:maxCount 1` _and_ the `sh:datatype xsd:string` constraints, instead of
just the `sh:maxCount` constraint.

dpetran committed:
remove useless thread macro

dpetran committed:
rename sparql-in to in

It's not sparql-specific, so no need to qualify it.

dpetran committed:
fix abs expression parsing

abs expects only a single expression as an argument, and that argument does not need to
be quoted.

dpetran committed:
no need to literal-quote numeric functions

dpetran committed:
Merge pull request #797 from fluree/feature/sparql-in-not-in

translate SPARQL IN/NOT IN expresssions into FQL

dpetran committed:
Merge pull request #796 from fluree/feature/sparql-exists-not-exists

translate SPARQL EXISTS/NOT EXISTS filters into FQL

dpetran committed:
Merge pull request #795 from fluree/feature/fql-in-not-in

add support for `in` expression

dpetran committed:
Merge pull request #784 from fluree/feature/fql-exists-not-exists

Feature/fql exists not exists

dpetran committed:
Merge pull request #780 from fluree/feature/sparql-from-named

Feature/sparql from named

dpetran committed:
Merge pull request #778 from fluree/feature/sparql-func

Feature/sparql func

dpetran committed:
Merge pull request #769 from fluree/feature/sparql-values

SPARQL VALUES patterns

dpetran commented:
The CI was stuck waiting for a container so I merged and ran CI locally. All commits are now on main.
